Original migration definition for showback

sergio-ocon opened the pull request:
This PR adds the migrations needed for the new showback.
The models and factories will be part of an independent gem

sergio-ocon commented:
@miq-bot add_label chargeback, enhancement, "sql migration" ,"gem changes"

sergio-ocon commented:
@aljesusg Initial migration with the models
@miq-bot add_label sql migration, gem_changes

sergio-ocon commented:
@miq-bot add_label gem changes

Fryguy commented on Gemfile:
```diff
@@ -56,6 +56,7 @@ gem "inifile",                        "~>3.0",         :require => false
 gem "jbuilder",                       "~>2.5.0" # For the REST API
 gem "manageiq-api-client",            "~>0.1.0",       :require => false
 gem "mime-types",                     "~>2.6.1",       :require => "mime/types/columnar"
+gem "money-rails",                    "~>1"
```
It would be nice if we could find a way to delay load this and the initializer until someone actually makes a chargeback call.  cc @jrafanie

jrafanie replied:
yeah, @sergio-ocon is there a single entry point to run the initializer and require money-rails?

I don't see any calls to MoneyRails in here outside of the initializer.  When is this needed?

sergio-ocon replied:
The monetize in the migration needs the money-rails required.

jrafanie replied:
Can the places that use it do the require?  We're trying to avoid requiring everything just booting the rails app.  I'm unsure where money-rails is used.  Autoloading it in `config/initializers/money.rb` is problematic because we're trying to extract dependencies to the code that uses it.  To be honest, I can't even tell what in this PR uses MoneyRails besides the initializer.  We should make it more obvious what's using it.

sergio-ocon replied:
I've moved it into required: false, and then I require it in the migration. If we are going to move all the models into a gem I will require it there

Fryguy commented on db/migrate/20170519153350_add_showback_models.rb:
```diff
@@ -0,0 +1,73 @@
+
+    create_table :showback_rates, :id => :bigserial, :force => :cascade do |t|
+      t.monetize   :fixed_rate
+      t.monetize   :variable_rate
```
Note to self, I'm going to need this new gem for the schema repo.

Fryguy commented on db/migrate/20170519153350_add_showback_models.rb:
```diff
@@ -0,0 +1,73 @@
+    create_table :showback_usage_types, :id => :bigserial, :force => :cascade do |t|
+      t.string     :category, :allow_nil => false
+      t.string     :description
+      t.string     :measure, :allow_nil => false
```
@sergio-ocon Please remove all of the db/schema.rb cruft: `:id => bigserial, :force => cascade, :allow_nil, :default`, etc.

sergio-ocon replied:
Deleting all. How do you guarantee that the default value is {}? It seems to be less efficient in the model than in the database itself.

Fryguy commented on db/migrate/20170519153350_add_showback_models.rb:
```diff
@@ -0,0 +1,73 @@
+    end
+
+    create_table :showback_charges, :id => :bigserial, :force => :cascade do |t|
+      t.monetize   :cost
```
What does monetize do?  Can we get away with just creating the columns manually instead of needing an extension library?

sergio-ocon replied:
The monetize gem add two columns and wrap them with .monetize, (monetize_cents, and monetize_currency).

Fryguy commented on db/migrate/20170519153350_add_showback_models.rb:
```diff
@@ -0,0 +1,73 @@
+      t.string     :measure, :allow_nil => false
+      t.text       :dimensions, :array => true, :allow_nil => false, :default => []
+      t.timestamp  :updated_at
+      t.timestamp  :created_at
```
Prefer `t.timestamps` over creating the two columns separately.

Fryguy commented on db/migrate/20170519153350_add_showback_models.rb:
```diff
@@ -0,0 +1,73 @@
+      t.timestamp  :start_time, :allow_nil => false  # When the event start
+      t.timestamp  :end_time,   :allow_nil => false  # When the event finish
+      t.belongs_to :resource,   :allow_nil => false, :type => :bigint, :polymorphic => true
+      t.json       :context,    :allow_nil => false, :default => {}
```
Why not `jsonb`?  From what I've read, there's basically never a reason to use json over jsonb.

sergio-ocon replied:
You are right, and I am using jsonb in other places. Updating

Fryguy commented on config/initializers/money.rb:
```diff
@@ -0,0 +1,92 @@
+                           present:     true, # column will be created
+                           null:        false, # other options will be treated as column options
+                           default:     0
+  }
```
Is this only necessary for migration purposes or does the runtime of MoneyRails need this as well?

sergio-ocon replied:
That is what monetize do, it gets this file and uses the configuration. It makes easier to maintain the gem in the code, as it generates everything automatically

sergio-ocon commented:
@aljesusg Can you check if we can use this migration in the provider?

lpichler commented on db/migrate/20170519153350_add_showback_models.rb:
```diff
@@ -0,0 +1,75 @@
+    end
+  end
+
+  def down
```
This is not needed when you will use `change` method instead of `up` method

lpichler replied:
@sergio-ocon Can you change it to `change` method? Thanks!

lpichler commented:
@sergio-ocon Do you have entity relationship diagram? For better understanding (I think Rubymine can generate it automatically) thanks!

Fryguy commented on Gemfile:
```diff
@@ -56,6 +56,7 @@ gem "inifile",                        "~>3.0",         :require => false
 gem "jbuilder",                       "~>2.5.0" # For the REST API
 gem "manageiq-api-client",            "~>0.1.0",       :require => false
 gem "mime-types",                     "~>2.6.1",       :require => "mime/types/columnar"
+gem "money-rails",                    "~>1",           :require  => false
```
I don't think this gem is needed in this PR anymore.

sergio-ocon commented:
Reviewed and updated.

@Fryguy, it is possible to set up fixtures and factories in a gem? We are having problems with that

sergio-ocon commented:
@miq-bot remove_label gem changes, wip

aljesusg commented:
@sergio-ocon,@Fryguy check in https://github.com/ManageIQ/manageiq/blob/master/spec/factories/physical_server.rb this model is defined in manageiq/master and not in lenovo provider.

The same for
https://github.com/ManageIQ/manageiq/blob/master/spec/factories/auth_key_pair_amazon.rb

sergio-ocon commented:
@Fryguy Let us know how do you want us to do this. We are losing plenty of our personal time to do this as you suggested and it is not done anywhere else in the code.

aljesusg commented:
@Fryguy Can we merge this and make another PR with the factories?

sergio-ocon commented:
That would make it easier to migrate the factories after

sergio-ocon commented:
@lpichler Can you review this?
Should we revert to the original commit with the model?
Please advise as we are blocked

sergio-ocon commented:
Hi @Fryguy If you merge this, then we will continue working to create the provider (or anything preeferred that works), in later PR

isimluk commented on db/migrate/20170519153350_add_showback_models.rb:
```diff
@@ -0,0 +1,75 @@
+      t.timestamps
+    end
+    add_index  :showback_events, :resource_id
+    add_index  :showback_events, :resource_type
```
Why do you prefer two indexes instead of single index with composite key?

sergio-ocon replied:
included in the belongs_to

sergio-ocon commented:
@isimluk Changed the migration with the index

Fryguy commented on db/migrate/20170519153350_add_showback_models.rb:
```diff
@@ -0,0 +1,73 @@
+      t.belongs_to :showback_price_plan, :type => :bigint
+      t.timestamps
+    end
+    add_index :showback_rates, :screener, :using => :gin
```
gin?  Why gin?

sergio-ocon replied:
It is jsonb because it is independent on the content. This is key element here: I don't care what is in chargeback, our rating is independent on anything that is inside. Using columns will make again the relationship manual, and thus I won't have what I want. With this schema I can charge anything (i.e. any data, any context, any tag, anything), with the same function. Adding a new charge is a matter of adding a new event and nothing else, I don't need to modify rating because I want to charge another event. If I do that with columns, then I need to do a migration every time I want to include a new event or system. (i.e. networking, storage, containers, pods, dockerfiles, you name it)

We don't want columns.
- We would need to define columns every time we want to rate, and that would not work.
- I want to use screener to match against context in a generic way: {screener} should be included in {context} to be used. I don't know before hand what is in screener or what is in context in the showback_event (i.e. tag: silver in context should have to be in screener to be used). More than that, I don0't care, that is for the customer to define, not me.
- In the way defined right now, it is quite easy to get that content into a non-sql database and use it directly.

Fryguy commented:
@sergio-ocon Finally, can you squash your commits here?  There are a lot of incidental commits that just clutter up the history.

sergio-ocon commented:
Done

gtanzillo commented on db/migrate/20170519153350_add_showback_models.rb:
```diff
@@ -0,0 +1,73 @@
+    end
+
+    create_table :showback_rates do |t|
+      t.bigint     :fixed_rate_cents # Columns needed by gem money
```
Do all these "cents" columns really need to be `bigint`?

sergio-ocon replied:
Int has a maximum value of  2_147_483_647, however, Money stores cents in that column, not ints, so the maximum is 21_474_836,47 $, it looks high, but it is not enough as we have some customers alreasy paying that amount on their monthly invoices (and everything yearly will be higher).
If you think that there are currencies that use 3 decimal digits, that gets reduced to 2M, and that is way too low.
For CCSP use cases, where you can have  multiple customers that you are serving, hardware management, etc. it gets worse.

jrafanie replied:
Ugh, I'm surprised it needs a specific column name to include a specific type: cents.  It might get confusing for other currencies that don't have cents.  A more generic term for the amount of the smallest denomination and with a separate column for the denomination type would probably be easier to understand.

Fryguy replied:
In @sergio-ocon 's defense, these particular column types are defined by the money gem, which is what he will be using, so they can't be changed / debated.

sergio-ocon commented:
Changed to subunits to avoid "cents"
Sqashed the commits to leave only one

sergio-ocon commented:
Hey @Fryguy, @chessbyte, @gtanzillo. This has been open for more than 17 days and we can't continue working until this is merged. Please merge it as it is.

We can always change whatever in a new PR.

martinpovolny commented:
@sergio-ocon : I sympatize but only a bit ;-)

There's a bunch of PRs that are far less conroversial and that are open much, much longer than 17 days.

Why can't you continue w/o this one merged?

martinpovolny commented:
@sergio-ocon : with all respect. I can see unresolved PR review feedback in this PR.

We have PRs that don't have any unresolved feedback and are waiting for review. That is bad and we need to improve that. But that is not the case of this PR.

I don't want to make you upset, but that is what I see here.

aljesusg commented:
LGFM :100:

sergio-ocon commented:
@martinpovolny I don't see any unresolved feedback, please point me to that. The only feedback that is pending is Jason comment on the index, and that is answered already as it shows a lack of knowledge on how this is working.

Do you see anything else

martinpovolny commented:
> https://github.com/ManageIQ/manageiq/pull/15165#discussion_r119651211

> @sergio-ocon Can you change it to change method? Thanks!

> https://github.com/ManageIQ/manageiq/pull/15165#issuecomment-303064319

> @sergio-ocon Do you have entity relationship diagram? For better understanding (I think Rubymine can generate it automatically) thanks!

@lpichler is one of the people who do a lot of work on chargeback (and @isimluk) I would suggest working with those people and responding their comments.

If you respond just to what @Fryguy wrote then you need @Fryguy to look into everything and merge everything. That is a bottleneck. We need to remove bottlenecks. I mean not removing @Fryguy! ;-) I mean elevating other people so that they will be able to do the job. (Such as becomming subject experts on chargeback).

isimluk commented:
This is possibly giant leap pr, so it is worth prudent discussion.

What are the plans for chargeback sub-system after this showback stuff is merged?

This is important! If this showback stuff is to replace current chargeback, it would be great to have technical RFC open for the public community review. That way we can collaborate over design and avoid any argumentation fallacies (if exists).

isimluk commented:
Also, I wonder, why can't we have migrations in the separate repos.

Long long time ago, when I introduced one of the foreman plug-ins. I was able to have custom migrations for my own classes in the new plug-in repo. Do we have public technical discussion (issue) why we don't allow that?

sergio-ocon commented:
Hey @isimluk:
- I can't generate a ER model. I can generate ManageIQ's and I can't export it. If I do that, my laptop freezes because I don't have enough memory. So the answer is no, I can't add an ER model into this. I talked to him on gitter to explain that.
- I've done 4 demos, and 1 code review already. We are scheduling the next one. Is a good suggestion.
- I would love to have the migrations in a separate repo. We don't do that because we have been told to keep migrations here. In fact, we need to put factories here because of that, so a second PR will come (in this PR I was told to remove them so I won't add them again to this PR)
- The migration is already changed into change from up. I did make the mistake of not deleting the down. Good catch!
- My rubocop is giving me a different error than the one used by this review. Is the gemfile ok?
- I've deleted the index as Jason requested anyway, I can add it later.
- I've changed bucket into pool, some people didn't like bucket.

That should address 100% of the requests, some more than once, some conflicting.

aljesusg commented:
Is this PR ok?

jrafanie commented:
> This PR adds the migrations needed for the new showback.
> The models will be part of an independent gem
> We also added the money_rails (MIT) gem to make sure that we support different currencies, conversion between them, calculations are done in integers, etc.

@sergio-ocon Can you review the description?  It seems like it was written when the PR originally had the money_rails gem in it.  It has been removed so the description needs to be updated to describe how the gem will interact with these schema changes.

> @sergio-ocon Finally, can you squash your commits here? There are a lot of incidental commits that just clutter up the history.

@sergio-ocon Please squash incidental commits unless they're worth keeping as independent commits.

isimluk commented:
> What are the plans for chargeback sub-system after this showback stuff is merged?

sergio-ocon commented:
1.Migration (this PR)
2. Basic factories (needed to make this work) --> in preparation
3. Gem. Models inside the gem. Porting provider to gem.
4. Cover 100% of existing functionality with models, except tiers, API and console only
5. Generate graph inventory and refresh for showback_events, automation of functions in steps before
6. Tiers
7. UI: assignments, rates, reports, currencies

Et voilà

aljesusg commented:
Thanks @Fryguy  !!! We can continue now
